File: Source/Server/Game/Objects/Animation.vb

Imports Mirage.Sharp.Asfw
Imports Core
Imports Newtonsoft.Json
Imports Newtonsoft.Json.Linq

Friend Module Animation

#Region "Database"
    Sub SaveAnimation(animationNum As Integer)
        Dim json As String = JsonConvert.SerializeObject(Type.Animation(animationNum)).ToString()

        If RowExists(animationNum, "animation")
            UpdateRow(animationNum, json, "animation", "data")
        Else
            InsertRow(animationNum, json, "animation")
        End If
    End Sub

    Sub LoadAnimations()
        Dim i As Integer

        For i = 1 To MAX_ANIMATIONS
            LoadAnimation(i)
        Next

    End Sub

    Sub LoadAnimation(animationNum As Integer)
        Dim data As JObject

        data = SelectRow(animationNum, "animation", "data")

        If data Is Nothing Then
            ClearAnimation(animationNum)
            Exit Sub
        End If

        Dim animationData = JObject.FromObject(data).toObject(Of AnimationStruct)()
        Type.Animation(animationNum) = animationData
    End Sub

    Sub ClearAnimation(index As Integer)
        Type.Animation(index).Name = ""
        Type.Animation(index).Sound = ""
        ReDim Type.Animation(index).Sprite(1)
        ReDim Type.Animation(index).Frames(1)
        ReDim Type.Animation(index).LoopCount(1)
        ReDim Type.Animation(index).LoopTime(1)
        Type.Animation(index).LoopCount(0) = 1
        Type.Animation(index).LoopCount(1) = 1
        Type.Animation(index).LoopTime(0) = 1
        Type.Animation(index).LoopTime(1) = 1
    End Sub

    Sub ClearAnimations()
         For i = 1 To MAX_ANIMATIONS
            ClearAnimation(i)
        Next
    End Sub

    Function AnimationData(AnimationNum As Integer) As Byte()
        Dim buffer As New ByteStream(4)

        buffer.WriteInt32(AnimationNum)
        For i = 0 To Type.Animation(AnimationNum).Frames.Length - 1
            buffer.WriteInt32(Type.Animation(AnimationNum).Frames(i))
        Next

        For i = 0 To Type.Animation(AnimationNum).LoopCount.Length - 1
            buffer.WriteInt32(Type.Animation(AnimationNum).LoopCount(i))
        Next

        For i = 0 To Type.Animation(AnimationNum).LoopTime.Length - 1
            buffer.WriteInt32(Type.Animation(AnimationNum).LoopTime(i))
        Next

        buffer.WriteString((Type.Animation(AnimationNum).Name))
        buffer.WriteString((Type.Animation(AnimationNum).Sound))

        For i = 0 To UBound(Type.Animation(AnimationNum).Sprite)
            buffer.WriteInt32(Type.Animation(AnimationNum).Sprite(i))
        Next

        Return buffer.ToArray
    End Function

#End Region

#Region "Incoming Packets"

    Sub Packet_EditAnimation(index As Integer, ByRef data() As Byte)
        ' Prevent hacking
        If GetPlayerAccess(index) < AccessType.Developer Then Exit Sub
        If TempPlayer(index).Editor > 0 Then Exit Sub

        Dim user As String

        user = IsEditorLocked(index, EditorType.Animation)

        If user <> "" Then 
            PlayerMsg(index, "The game editor is locked and being used by " + user + ".", ColorType.BrightRed)
            Exit Sub
        End If

        TempPlayer(index).Editor = EditorType.Animation

        SendAnimations(index)

        Dim Buffer = New ByteStream(4)
        Buffer.WriteInt32(ServerPackets.SAnimationEditor)
        Socket.SendDataTo(index, Buffer.Data, Buffer.Head)
        Buffer.Dispose()
    End Sub

    Sub Packet_SaveAnimation(index As Integer, ByRef data() As Byte)
        Dim AnimNum As Integer
        Dim buffer As New ByteStream(data)

        AnimNum = buffer.ReadInt32

        ' Update the Animation
        For i = 0 To UBound(Type.Animation(AnimNum).Frames)
            Type.Animation(AnimNum).Frames(i) = buffer.ReadInt32()
        Next

        For i = 0 To UBound(Type.Animation(AnimNum).LoopCount)
            Type.Animation(AnimNum).LoopCount(i) = buffer.ReadInt32()
        Next

        For i = 0 To UBound(Type.Animation(AnimNum).LoopTime)
            Type.Animation(AnimNum).LoopTime(i) = buffer.ReadInt32()
        Next

        Type.Animation(AnimNum).Name = buffer.ReadString()
        Type.Animation(AnimNum).Sound = buffer.ReadString()

        For i = 0 To UBound(Type.Animation(AnimNum).Sprite)
            Type.Animation(AnimNum).Sprite(i) = buffer.ReadInt32()
        Next

        buffer.Dispose()

        ' Save it
        SaveAnimation(AnimNum)
        SendUpdateAnimationToAll(AnimNum)
        Addlog(GetPlayerLogin(index) & " saved Animation #" & AnimNum & ".", ADMIN_LOG)

    End Sub

    Sub Packet_RequestAnimation(index As Integer, ByRef data() As Byte)
        Dim Buffer = New ByteStream(data), n As Integer

        n = Buffer.ReadInt32

        If n <= 0 Or n > MAX_ANIMATIONS Then Exit Sub

        SendUpdateAnimationTo(index, n)
    End Sub

#End Region

#Region "Outgoing Packets"

    Sub SendAnimation(MapNum As Integer, Anim As Integer, X As Integer, Y As Integer, Optional LockType As Byte = 0, Optional Lockindex As Integer = 0)
        Dim buffer As New ByteStream(4)
        buffer.WriteInt32(ServerPackets.SAnimation)
        buffer.WriteInt32(Anim)
        buffer.WriteInt32(X)
        buffer.WriteInt32(Y)
        buffer.WriteInt32(LockType)
        buffer.WriteInt32(Lockindex)

        SendDataToMap(MapNum, buffer.Data, buffer.Head)

        buffer.Dispose()
    End Sub

    Sub SendAnimations(index As Integer)
        Dim i As Integer

       For i = 1 To MAX_ANIMATIONS

            If Len(Type.Animation(i).Name) > 0 Then
                SendUpdateAnimationTo(index, i)
            End If

        Next

    End Sub

    Sub SendUpdateAnimationTo(index As Integer, AnimationNum As Integer)
        Dim buffer As New ByteStream(4)

        buffer.WriteInt32(ServerPackets.SUpdateAnimation)

        buffer.WriteBlock(AnimationData(AnimationNum))

        Socket.SendDataTo(index, buffer.Data, buffer.Head)
        buffer.Dispose()
    End Sub

    Sub SendUpdateAnimationToAll(AnimationNum As Integer)
        Dim buffer As New ByteStream(4)

        buffer.WriteInt32(ServerPackets.SUpdateAnimation)

        buffer.WriteBlock(AnimationData(AnimationNum))

        SendDataToAll(buffer.Data, buffer.Head)
        buffer.Dispose()
    End Sub

#End Region

End Module

File: Source/Server/Game/Objects/Animation.cs
using System;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;
using Mirage.Sharp.Asfw;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using static Core.Packets;
using static Core.Type;
using static Core.Global.Command;
using System.Threading.Tasks;

namespace Server
{

    public class Animation
    {

        #region Database
        public static void SaveAnimation(int animationNum)
        {
            string json = JsonConvert.SerializeObject(Core.Data.Animation[animationNum]).ToString();

            if (Database.RowExists(animationNum, "animation"))
            {
                Database.UpdateRow(animationNum, json, "animation", "data");
            }
            else
            {
                Database.InsertRow(animationNum, json, "animation");
            }
        }

        public static async System.Threading.Tasks.Task LoadAnimationsAsync()
        {
            int i;
            var loopTo = Core.Constant.MAX_ANIMATIONS - 1;
            for (i = 0; i < loopTo; i++)
                await System.Threading.Tasks.Task.Run(() => LoadAnimation(i));
        }

        public static async System.Threading.Tasks.Task LoadAnimation(int animationNum)
        {
            JObject data;

            data = await Database.SelectRowAsync(animationNum, "animation", "data");

            if (data is null)
            {
                ClearAnimation(animationNum);
                return;
            }

            var animationData = JObject.FromObject(data).ToObject<Core.Type.Animation>();
            Core.Data.Animation[animationNum] = animationData;
        }

        public static void ClearAnimation(int index)
        {
            Core.Data.Animation[index].Name = "";
            Core.Data.Animation[index].Sound = "";
            Core.Data.Animation[index].Sprite = new int[2];
            Core.Data.Animation[index].Frames = new int[2];
            Core.Data.Animation[index].LoopCount = new int[2];
            Core.Data.Animation[index].LoopTime = new int[2];
            Core.Data.Animation[index].LoopCount[0] = 0;
            Core.Data.Animation[index].LoopCount[1] = 0;
            Core.Data.Animation[index].LoopTime[0] = 0;
            Core.Data.Animation[index].LoopTime[1] = 0;
        }

        public static void ClearAnimations()
        {
            for (int i = 0, loopTo = Core.Constant.MAX_ANIMATIONS; i < loopTo; i++)
                ClearAnimation(i);
        }

        public static byte[] AnimationData(int AnimationNum)
        {
            var buffer = new ByteStream(4);

            buffer.WriteInt32(AnimationNum);
            for (int i = 0, loopTo = Information.UBound(Core.Data.Animation[AnimationNum].Frames); i < loopTo; i++)
                buffer.WriteInt32(Core.Data.Animation[AnimationNum].Frames[i]);

            for (int i = 0, loopTo1 = Information.UBound(Core.Data.Animation[AnimationNum].LoopCount); i < loopTo1; i++)
                buffer.WriteInt32(Core.Data.Animation[AnimationNum].LoopCount[i]);

            for (int i = 0, loopTo2 = Information.UBound(Core.Data.Animation[AnimationNum].LoopTime); i < loopTo2; i++)
                buffer.WriteInt32(Core.Data.Animation[AnimationNum].LoopTime[i]);

            buffer.WriteString(Core.Data.Animation[AnimationNum].Name);
            buffer.WriteString(Core.Data.Animation[AnimationNum].Sound);

            for (int i = 0, loopTo3 = Information.UBound(Core.Data.Animation[AnimationNum].Sprite); i < loopTo3; i++)
                buffer.WriteInt32(Core.Data.Animation[AnimationNum].Sprite[i]);

            return buffer.ToArray();
        }

        #endregion

        #region Incoming Packets

        public static void Packet_RequestEditAnimation(int index, ref byte[] data)
        {
            // Prevent hacking
            if (GetPlayerAccess(index) < (byte) Core.AccessLevel.Developer)
                return;

            string user;

            user = IsEditorLocked(index, (byte) Core.EditorType.Animation);

            if (!string.IsNullOrEmpty(user))
            {
                NetworkSend.PlayerMsg(index, "The game editor is locked and being used by " + user + ".", (int) Core.Color.BrightRed);
                return;
            }

            Core.Data.TempPlayer[index].Editor = (byte) Core.EditorType.Animation;

            SendAnimations(index);

            var buffer = new ByteStream(4);
            buffer.WriteInt32((int) ServerPackets.SAnimationEditor);
            NetworkConfig.Socket.SendDataTo(index, buffer.UnreadData, buffer.WritePosition);
            buffer.Dispose();
        }

        public static void Packet_SaveAnimation(int index, ref byte[] data)
        {
            int AnimNum;
            var buffer = new ByteStream(data);

            AnimNum = buffer.ReadInt32();

            // Update the Animation
            for (int i = 0, loopTo = Information.UBound(Core.Data.Animation[AnimNum].Frames); i < loopTo; i++)
                Core.Data.Animation[AnimNum].Frames[i] = buffer.ReadInt32();

            for (int i = 0, loopTo1 = Information.UBound(Core.Data.Animation[AnimNum].LoopCount); i < loopTo1; i++)
                Core.Data.Animation[AnimNum].LoopCount[i] = buffer.ReadInt32();

            for (int i = 0, loopTo2 = Information.UBound(Core.Data.Animation[AnimNum].LoopTime); i < loopTo2; i++)
                Core.Data.Animation[AnimNum].LoopTime[i] = buffer.ReadInt32();

            Core.Data.Animation[AnimNum].Name = buffer.ReadString();
            Core.Data.Animation[AnimNum].Sound = buffer.ReadString();

            for (int i = 0, loopTo3 = Information.UBound(Core.Data.Animation[AnimNum].Sprite); i < loopTo3; i++)
                Core.Data.Animation[AnimNum].Sprite[i] = buffer.ReadInt32();

            buffer.Dispose();

            // Save it
            SaveAnimation(AnimNum);
            SendUpdateAnimationToAll(AnimNum);
            Core.Log.Add(GetAccountLogin(index) + " saved Animation #" + AnimNum + ".", Constant.ADMIN_LOG);

        }

        public static void Packet_RequestAnimation(int index, ref byte[] data)
        {
            var buffer = new ByteStream(data);
            int n;

            n = buffer.ReadInt32();

            if (n < 0 | n > Core.Constant.MAX_ANIMATIONS)
                return;

            SendUpdateAnimationTo(index, n);
        }

        #endregion

        #region Outgoing Packets

        public static void SendAnimation(int mapNum, int Anim, int X, int Y, byte LockType = 0, int Lockindex = 0)
        {
            var buffer = new ByteStream(4);
            buffer.WriteInt32((int) ServerPackets.SAnimation);
            buffer.WriteInt32(Anim);
            buffer.WriteInt32(X);
            buffer.WriteInt32(Y);
            buffer.WriteInt32(LockType);
            buffer.WriteInt32(Lockindex);

            NetworkConfig.SendDataToMap(mapNum, buffer.UnreadData, buffer.WritePosition);

            buffer.Dispose();
        }

        public static void SendAnimations(int index)
        {
            int i;

            var loopTo = Core.Constant.MAX_ANIMATIONS - 1;
            for (i = 0; i < loopTo; i++)
            {

                if (Strings.Len(Core.Data.Animation[i].Name) > 0)
                {
                    SendUpdateAnimationTo(index, i);
                }

            }

        }

        public static void SendUpdateAnimationTo(int index, int AnimationNum)
        {
            var buffer = new ByteStream(4);

            buffer.WriteInt32((int) ServerPackets.SUpdateAnimation);

            buffer.WriteBlock(AnimationData(AnimationNum));

            NetworkConfig.Socket.SendDataTo(index, buffer.UnreadData, buffer.WritePosition);
            buffer.Dispose();
        }

        public static void SendUpdateAnimationToAll(int AnimationNum)
        {
            var buffer = new ByteStream(4);

            buffer.WriteInt32((int) ServerPackets.SUpdateAnimation);

            buffer.WriteBlock(AnimationData(AnimationNum));

            NetworkConfig.SendDataToAll(buffer.UnreadData, buffer.WritePosition);
            buffer.Dispose();
        }

        #endregion

    }
}